e3nn_jax/_src/spherical_signal.py

```diff
@@ -0,0 +1,215 @@
+        resampled_grid_values = e3nn.to_s2grid(coeffs, *grid_resolution, quadrature=self.quadrature)
+        return SphericalSignal(
+            resampled_grid_values, self.y, self.alpha, self.quadrature, self.quadrature_weights, self.p_val, self.p_arg
+        )
+
+    def _coeffs(self, lmax: int) -> e3nn.IrrepsArray:
+        """Returns the coefficients for the spherical harmonics upto l = lmax."""
+        irreps = SphericalSignal.irreps_for_spherical_signal(lmax, p_val=self.p_val, p_arg=self.p_arg)
+        return e3nn.from_s2grid(self.grid_values, irreps, quadrature=self.quadrature)
+
+    def transform_by_angles(self, alpha: float, beta: float, gamma: float, lmax: int, k: int = 0) -> "SphericalSignal":
+        """Rotate the signal by the given Euler angles."""
+        coeffs = self._coeffs(lmax)
+        transformed_coeffs = coeffs.transform_by_angles(alpha, beta, gamma)
+        transformed_grid_values = e3nn.to_s2grid(transformed_coeffs, *self.grid_resolution(), quadrature=self.quadrature)
+        return SphericalSignal(
+            transformed_grid_values, self.y, self.alpha, self.quadrature, self.quadrature_weights, self.p_val, self.p_arg
+        )
+
+    def transform_by_matrix(self, R: chex.Array, lmax: int) -> "SphericalSignal":
+        """Rotate the signal by the given rotation matrix."""
+        coeffs = self._coeffs(lmax)
+        transformed_coeffs = coeffs.transform_by_matrix(R)
+        transformed_grid_values = e3nn.to_s2grid(transformed_coeffs, *self.grid_resolution(), quadrature=self.quadrature)
+        return SphericalSignal(
+            transformed_grid_values, self.y, self.alpha, self.quadrature, self.quadrature_weights, self.p_val, self.p_arg
+        )
+
+    def transform_by_axis_angle(self, axis: chex.Array, angle: float, lmax: int, k: int = 0) -> "SphericalSignal":
+        """Rotate the signal by the given angle around an axis."""
+        coeffs = self._coeffs(lmax)
+        transformed_coeffs = coeffs.transform_by_axis_angle(axis, angle, k)
+        transformed_grid_values = e3nn.to_s2grid(transformed_coeffs, *self.grid_resolution(), quadrature=self.quadrature)
+        return SphericalSignal(
+            transformed_grid_values, self.y, self.alpha, self.quadrature, self.quadrature_weights, self.p_val, self.p_arg
+        )
+
+    def transform_by_quaternion(self, q: chex.Array, lmax: int, k: int = 0) -> "SphericalSignal":
+        """Rotate the signal by the given quaternion."""
+        coeffs = self._coeffs(lmax)
+        transformed_coeffs = coeffs.transform_by_quaternion(q, k)
+        transformed_grid_values = e3nn.to_s2grid(transformed_coeffs, *self.grid_resolution(), quadrature=self.quadrature)
+        return SphericalSignal(
+            transformed_grid_values, self.y, self.alpha, self.quadrature, self.quadrature_weights, self.p_val, self.p_arg
+        )
+
+    def apply(self, func: Callable[[chex.Array], chex.Array]):
+        """Applies a function pointwise on the grid."""
```

Comment:
Use `jax.eval_shape` to check that the shape is unaffected and raise an error otherwise